### PR TITLE
Add ALIGN to transformers

### PR DESCRIPTION
**Pull request by alaradirik**
# What does this PR do?
Adds [ALIGN](https://arxiv.org/abs/2102.05918) to transformers, a multi-modal model similar to CLIP. ALIGN uses EfficientNet as its vision encoder and BERT as its text encoder. No public implementation is available for this model, the code is adapted from Kakao Brain's tensorflow implementation shared with us.

To do:
- [x] Upload converted model
- [x] Create a model card
- [x] Fix processor tests
- [x] Fix model tests

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [X ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [X ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X ] Did you write any new necessary tests?

### CODE REVIEW

File: src/transformers/models/align/modeling_align.py [hunk 1]
```diff
@@ -0,0 +1,1082 @@
...
+    def forward(self, hidden_states: torch.FloatTensor) -> torch.Tensor:
+        # Expand phase
+        hidden_states = self.expand_conv(hidden_states)
+        hidden_states = self.expand_bn(hidden_states)
+        hidden_states = self.expand_act(hidden_states)
+
+        return hidden_states
+
+
+class ALIGNDepthwiseLayer(nn.Module):
```
Comment (amyeroberts) [inline src/transformers/models/align/modeling_align.py:408]: same here re same as in efficientnet

File: src/transformers/models/align/modeling_align.py [hunk 2]
```diff
@@ -0,0 +1,1082 @@
...
+            hidden_states = self.depthwise_conv_pad(hidden_states)
+
+        hidden_states = self.depthwise_conv(hidden_states)
+        hidden_states = self.depthwise_norm(hidden_states)
+        hidden_states = self.depthwise_act(hidden_states)
+
+        return hidden_states
+
+
+class ALIGNSqueezeExciteLayer(nn.Module):
```
Comment (amyeroberts) [inline src/transformers/models/align/modeling_align.py:447]: same here re same as in efficientnet

File: src/transformers/models/align/modeling_align.py [hunk 3]
```diff
@@ -0,0 +1,1082 @@
...
+        hidden_states = self.act_reduce(hidden_states)
+
+        hidden_states = self.expand(hidden_states)
+        hidden_states = self.act_expand(hidden_states)
+        hidden_states = torch.mul(inputs, hidden_states)
+
+        return hidden_states
+
+
+class ALIGNFinalBlockLayer(nn.Module):
```
Comment (amyeroberts) [inline src/transformers/models/align/modeling_align.py:486]: same here re same as in efficientnet

File: src/transformers/models/align/modeling_align.py [hunk 4]
```diff
@@ -0,0 +1,1082 @@
...
+            hidden_states = self.expansion(hidden_states)
+        hidden_states = self.depthwise_conv(hidden_states)
+
+        # Squeeze and excite phase
+        hidden_states = self.squeeze_excite(hidden_states)
+        hidden_states = self.projection(embeddings, hidden_states)
+        return hidden_states
+
+
+class ALIGNVisionEncoder(nn.Module):
```
Comment (amyeroberts) [inline src/transformers/models/align/modeling_align.py:598]: same here re same as in efficientnet

File: src/transformers/models/align/modeling_align.py [hunk 5]
```diff
@@ -0,0 +1,1082 @@
...
+@add_start_docstrings(
+    """The text model from ALIGN without any head or projection on top.""",
+    ALIGN_START_DOCSTRING,
+)
+class ALIGNTextModel(ALIGNPreTrainedModel):
+    config_class = ALIGNTextConfig
+
+    def __init__(self, config: ALIGNTextConfig):
+        super().__init__(config)
+        self.text_model = BertModel(config)
```
Comment (amyeroberts) [inline src/transformers/models/align/modeling_align.py:712]: Please use `AutoModel` here instead of importing a specific model class

File: src/transformers/models/align/processing_align.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            - **attention_mask** -- List of indices specifying which tokens should be attended to by the model (when
+              `return_attention_mask=True` or if *"attention_mask"* is in `self.model_input_names` and if `text` is not
+              `None`).
+            - **pixel_values** -- Pixel values to be fed to a model. Returned when `images` is not `None`.
+        """
+        if text is None and images is None:
+            raise ValueError("You have to specify either text or images. Both cannot be none.")
+
+        if text is not None:
+            encoding = self.tokenizer(text, return_tensors=return_tensors, **kwargs)
```
Comment (amyeroberts) [inline src/transformers/models/align/processing_align.py:87]: Can you make sure to pass specific kwargs to the respective tokenizer and image processer e.g. [like this in FLAVA](https://github.com/huggingface/transformers/blob/92dfceb12446404fe547b1271ecaf86d2498546b/src/transformers/models/flava/processing_flava.py#L94)

File: src/transformers/models/align/processing_align.py [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+        if text is not None and images is not None:
+            encoding["pixel_values"] = image_features.pixel_values
+            return encoding
+        elif text is not None:
+            return encoding
+        else:
+            return BatchEncoding(data=dict(**image_features), tensor_type=return_tensors)
+
+    def adjust_sequence_length(self, encoding, max_seq_len, return_tensors):
```
Comment (amyeroberts) [inline src/transformers/models/align/processing_align.py:102]: Is this needed? As far as I can tell, this is just truncating too long sequences and padding to the max length, which tokenizers can already do for you e.g.:

```
tokenizer(
    ["this is a very long string that will be truncated", "hi"], 
    truncation=True, 
    padding='max_length', 
    max_length=5
)
```

File: src/transformers/models/efficientnet/image_processing_efficientnet.py [hunk 8]
```diff
@@ -328,13 +328,10 @@ def preprocess(
 
         if do_resize:
             images = [self.resize(image=image, size=size, resample=resample) for image in images]
-
```
Comment (amyeroberts) [inline src/transformers/models/efficientnet/image_processing_efficientnet.py:331]: Why remove the spaces here?

File: tests/models/align/test_processor_align.py [hunk 9]
```diff
@@ -0,0 +1,207 @@
+# Copyright 2021 The HuggingFace Team. All rights reserved.
```
Comment (amyeroberts) [inline tests/models/align/test_processor_align.py:1]:
```suggestion
# Copyright 2023 The HuggingFace Team. All rights reserved.
```

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this model! Most comments are just formatting nits. 

Two main comments to address:
- Can you modify the model prefix to Align, rather than ALIGN?
- All of the vision encoder architecture is, as far as I can tell, exactly the same as the EfficientNet model. And the text encoder is directly imported from Bert. Is there a reason for reimplementing EfficientNet but importing Bert?

**Comment by alaradirik** [comment]
> Thanks for adding this model! Most comments are just formatting nits.
> 
> Two main comments to address:
> 
> * Can you modify the model prefix to Align, rather than ALIGN?
> * All of the vision encoder architecture is, as far as I can tell, exactly the same as the EfficientNet model. And the text encoder is directly imported from Bert. Is there a reason for reimplementing EfficientNet but importing Bert?

Thank you! I will modify the model prefix to Align but the Kakao Brain EfficientNet implementation is slightly different from the official one (final layers), hence I'm copy pasting the modules and changing the final layers.

File: src/transformers/models/align/modeling_align.py [hunk 10]
```diff
@@ -0,0 +1,1082 @@
...
+    add_start_docstrings_to_model_forward,
+    logging,
+    replace_return_docstrings,
+)
+from .configuration_align import ALIGNConfig, ALIGNTextConfig, ALIGNVisionConfig
+
+
+logger = logging.get_logger(__name__)
+
+_CHECKPOINT_FOR_DOC = "kakaobrain/align-base"
```
[the remark this answers is not in the document]
Comment (alaradirik) [reply src/transformers/models/align/modeling_align.py:45]: The ALIGN code is not open-source, the code and model weights are provided by the Kakao Brain team

File: src/transformers/models/align/processing_align.py (hunk 7 above)
Comment (alaradirik) [reply src/transformers/models/align/processing_align.py:102]: Makes sense! I'm removing `adjust_sequence_length()`

File: src/transformers/models/efficientnet/image_processing_efficientnet.py (hunk 8 above)
Comment (alaradirik) [reply src/transformers/models/efficientnet/image_processing_efficientnet.py:331]: Readding them

**Comment by amyeroberts** [comment]
> The Kakao Brain EfficientNet implementation is slightly different from the official one (final layers), hence I'm copy pasting the modules and changing the final layers.

OK, I understand better now. I realised there's something I overlooked in my initial review: why are we using a randomly initialized Bert model as the text encoder? My understanding of this model is a text and vision backbone are loaded in and then trained to align their embeddings i.e. the weights we should be loading are the respective BERT and EfficientNet weights post-training and the api would be something like `AlignModel.from_pretrained(google/align-efficientnetv2-bert-large')`. Is this correct?

File: src/transformers/models/align/modeling_align.py (hunk 5 above)
Comment (alaradirik) [reply src/transformers/models/align/modeling_align.py:715]: ALIGNTextConfig is copy pasted from BertConfig, should I just initialise it as an instance of BertConfig similar to how SwinConfig is used in Mask2Former?

**Comment by alaradirik** [comment]
> I can modify the model prefix to Align but then I will have to change README titles to Align in order to not get repo-consistency errors and I'd like to keep the model name in the documentation the same as the original work. Could we keep it as it is, similar to BERT and CLIP? cc @sgugger
> 
My mistake, I figured out how to keep the documentation title camel cased, disregard this please.

File: src/transformers/models/align/modeling_align.py (hunk 4 above)
Comment (alaradirik) [reply src/transformers/models/align/modeling_align.py]: This module (KakaoBrain's EfficientNet backbone for ALIGN) is slightly different from the official EfficientNet implementation we added

**Comment by amyeroberts** [comment]
Ah, OK. Sorry for the confusion. Just to make sure I understand: Google haven't released their weights; kakao Brain have themselves trained the ALIGN model and these are the weights we're using. Is this right?

File: src/transformers/models/align/modeling_align.py [hunk 11]
```diff
@@ -0,0 +1,1085 @@
...
+            if module.bias is not None:
+                module.bias.data.zero_()
+        elif isinstance(module, AlignModel):
+            nn.init.xavier_uniform_(module.text_projection.weight)
+
+        if isinstance(module, nn.LayerNorm):
+            module.bias.data.zero_()
+            module.weight.data.fill_(1.0)
+        if isinstance(module, nn.Linear) and module.bias is not None:
+            module.bias.data.zero_()
```
Comment (sgugger) [inline src/transformers/models/align/modeling_align.py:699]: This is already done above.
```suggestion
```

File: src/transformers/models/align/modeling_align.py (hunk 5 above)
Comment (sgugger) [reply src/transformers/models/align/modeling_align.py:715]: As said above, you should copy the BERT model here and adapt the prefixes to AlignText

**Review by sgugger — COMMENTED** [review]
Thanks for working on this new model. Left a couple of comments, the main one being on making sure to leave the model folders independent from each other. Like you copied the vision model code from EfficientNet, the text model code should be copied from BERT.

**Comment by alaradirik** [comment]
> Ah, OK. Sorry for the confusion. Just to make sure I understand: Google haven't released their weights; kakao Brain have themselves trained the ALIGN model and these are the weights we're using. Is this right?

No problem at all, Kakao Brain implemented and trained the ALIGN model themselves, Google haven't released checkpoints nor the code.

File: src/transformers/models/align/convert_align_tf_to_hf.py [hunk 12]
```diff
@@ -0,0 +1,388 @@
...
+
+
+@torch.no_grad()
+def convert_align_checkpoint(checkpoint_path, pytorch_dump_folder_path, save_model, push_to_hub):
+    """
+    Copy/paste/tweak model's weights to our ALIGN structure.
+    """
+    # Load original model
+    seq_length = 64
+    tok = Tokenizer(seq_length)
```
[the remark this answers is not in the document]
Comment (alaradirik) [reply src/transformers/models/align/convert_align_tf_to_hf.py:294]: This is to ensure the original implementation's preprocessor output matches ours but I can remove it, I already checked multiple times that they match

File: src/transformers/models/align/modeling_align.py [hunk 13]
```diff
@@ -0,0 +1,1085 @@
...
+from torch import nn
+
+from ...activations import ACT2FN
+from ...modeling_outputs import (
+    BaseModelOutputWithNoAttention,
+    BaseModelOutputWithPoolingAndCrossAttentions,
+    BaseModelOutputWithPoolingAndNoAttention,
+)
+from ...modeling_utils import PreTrainedModel
+from ...models.bert import BertModel
```
[the remark this answers is not in the document]
Comment (alaradirik) [reply src/transformers/models/align/modeling_align.py:32]: Got it, I will copy paste it as AlignTextModel

File: src/transformers/models/align/convert_align_tf_to_hf.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/models/align/convert_align_tf_to_hf.py:294]: No strong opinion if that works for you. Just weird to use that base class :-)

**Review by sgugger — APPROVED** [review]
Thanks a lot, LGTM! @amyeroberts do you want to have a final look as well?

**Review by amyeroberts — APPROVED** [review]
Thanks for adding this model and work on the PR!

Looks good to me - mostly nit comments. Just one question I have on L514 in `modeling_align.py` that should be addressed before merging.

File: src/transformers/models/align/modeling_align.py [hunk 14]
```diff
@@ -0,0 +1,1082 @@
+# coding=utf-8
+# Copyright 2023 The OpenAI Team Authors and The HuggingFace Team. All rights reserved.
```
[the remark this answers is not in the document]
Comment (amyeroberts) [reply src/transformers/models/align/modeling_align.py]: Good question. Yes, I'd say to include Kakao Brain since its their implementation we're following :)

File: docs/source/en/model_doc/align.mdx [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+specific language governing permissions and limitations under the License.
+-->
+
+# ALIGN
+
+## Overview
+
+The ALIGN model was proposed in [Scaling Up Visual and Vision-Language Representation Learning With Noisy Text Supervision](https://arxiv.org/abs/2102.05918) by Chao Jia, Yinfei Yang, Ye Xia, Yi-Ting Chen, Zarana Parekh, Hieu Pham, Quoc V. Le, Yunhsuan Sung, Zhen Li, Tom Duerig. ALIGN features a dual-encoder architecture with [EfficientNet](efficientnet) as its vision encoder and [BERT] as its text encoder, and learns to align visual and text representations with contrastive learning. Unlike previous work, ALIGN leverages a massive noisy dataset and shows that  the scale of the corpus can be used to achieve SOTA representations with a simple recipe.
```
[the remark this answers is not in the document]
Comment (alaradirik) [reply docs/source/en/model_doc/align.mdx:17]: Thanks, I'm fixing it now!

File: docs/source/en/model_doc/align.mdx [hunk 16]
```diff
@@ -0,0 +1,59 @@
...
+
+This model was contributed by [Alara Dirik](https://huggingface.co/adirik).
+The original code can is not released, this implementation is based on the Kakao Brain implementation based on the original paper.
```
[the remark this answers is not in the document]
Comment (alaradirik) [reply docs/source/en/model_doc/align.mdx:24]: They haven't open sourced their implementation, the inference code is in a private repo

File: src/transformers/models/align/modeling_align.py [hunk 17]
```diff
@@ -0,0 +1,1640 @@
...
+        )
+        self.dropout = nn.Dropout(p=drop_rate)
+
+    def forward(self, embeddings: torch.FloatTensor, hidden_states: torch.FloatTensor) -> torch.Tensor:
+        hidden_states = self.project_conv(hidden_states)
+        hidden_states = self.project_bn(hidden_states)
+
+        if self.apply_dropout:
+            hidden_states = self.dropout(hidden_states)
+            hidden_states = hidden_states + embeddings
```
[the remark this answers is not in the document]
Comment (alaradirik) [reply src/transformers/models/align/modeling_align.py:514]: Yes, it is the same in the original EfficientNet implementation
